### PR TITLE
feat(manager/pub): update to only changed dependencies

### LINKED ISSUE #19653: Keep changes in pubspec.lock to minimum
### What would you like Renovate to be able to do?

For a Dart (flutter) project I'd like to be able to configure renovate such that the lock file (pubspec.lock) has versions updated for related dependencies only. 

E.g. a pull request that updates package _A_ `1.7.0` -> `2.1.0` changes pubspec.yaml to specify `A: ^2.1.0` in place of `A: ^1.0.0` and changes pubspec.lock to specify
* `2.1.0` in place of `1.7.0` for _A_ and
* `3.1.5` in place of `3.0.2` for _B_ because _A@2.1.0_ specifies `B: ^3.1.0`.

Basically, the new lock file seems to be generated by running `flutter pub upgrade`, but I'd need it to be generated by running `flutter pub upgrade A`. That keeps changes in the lock file to a minimum.

I'm also using renovate for a nodejs project, which already behaves like this. That _and_ the flutter project just extend the preset `config:base` with some package rules. The only difference is that the nodejs project uses pinned dependencies while the flutter one does not.

### If you have any ideas on how this should be implemented, please tell us here.

lockFileMaintenance? range strategy? (as per https://github.com/renovatebot/renovate/discussions/19608)

I'm unsure because the nodejs project worked like that out-of-the-box.

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by zeshuaro**
## Changes

- Update to specify the updated dependencies for the `pub` manager when we attempt to lock their versions with `pub upgrade`
- Add more logging when updating the lock file
- I loosely referenced the code from the `poetry` manager

## Context

Closes #19653 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository
  - See [this](https://github.com/barriot/r1/pull/3/files) PR for the current behaviour, where transitive dependency in the lock file is updated as part of the main dependency update
  - See [this](https://github.com/barriot/r0/pull/8/files) PR for the new behaviour where the transitive dependency is not updated, only the main dependency is

### CODE REVIEW

File: lib/modules/manager/pub/artifacts.ts [hunk 1]
```diff
@@ -26,32 +29,45 @@ export async function updateArtifacts({
...
-    !(
-      config.updateType === 'lockFileMaintenance' ||
-      config.updateType === 'lockfileUpdate'
-    ) &&
-    updatedDeps.length < 1
-  ) {
+  logger.debug(`pub.updateArtifacts(${packageFileName})`);
+  const isLockFileMaintenance = config.updateType === 'lockFileMaintenance';
+
+  if (!is.nonEmptyArray(updatedDeps) && !isLockFileMaintenance) {
```
Comment (zeshuaro) [inline lib/modules/manager/pub/artifacts.ts:35]: Doesn't seem like any other managers are checking `config.updateType === 'lockfileUpdate'`, and I referenced other managers with this change

File: lib/modules/manager/pub/artifacts.ts [hunk 2]
```diff
@@ -60,16 +76,15 @@ export async function updateArtifacts({
...
         },
       ],
-    });
+    };
 
+    await exec(cmd, execOptions);
     const newLockFileContent = await readLocalFile(lockFileName, 'utf8');
-    if (
-      oldLockFileContent === newLockFileContent ||
-      newLockFileContent === undefined
```
Comment (zeshuaro) [inline lib/modules/manager/pub/artifacts.ts:68]: Doesn't seem like any other managers are checking `newLockFileContent === undefined` so I removed it

File: lib/modules/manager/pub/artifacts.ts (hunk 1 above)
Comment (JamieMagee) [inline lib/modules/manager/pub/artifacts.ts:35]: Avoids a double negative:
```suggestion
  if (is.emptyArray(updatedDeps) && !isLockFileMaintenance) {
```

**Comment by JamieMagee** [comment]
Does `pub upgrade` with a specific dependency also handle indirect dependencies? For example, if I depend on A, and A depends on B, does `pub upgrade A` also upgrade B when required?

**Comment by zeshuaro** [comment]
> Does pub upgrade with a specific dependency also handle indirect dependencies? For example, if I depend on A, and A depends on B, does pub upgrade A also upgrade B when required?

Yes it does, see this PR for example: https://github.com/barriot/r2/pull/12

File: lib/modules/manager/pub/artifacts.ts [hunk 3]
```diff
@@ -26,32 +29,45 @@ export async function updateArtifacts({
...
   const lockFileName = getSiblingFileName(packageFileName, 'pubspec.lock');
   const oldLockFileContent = await readLocalFile(lockFileName, 'utf8');
   if (!oldLockFileContent) {
+    logger.debug('No pubspec.lock found');
     return null;
   }
 
   try {
     await writeLocalFile(packageFileName, newPackageFileContent);
+    logger.debug(`Updating ${lockFileName}`);
```
Comment (viceice) [inline lib/modules/manager/pub/artifacts.ts:49]:
```suggestion
```
you'll see the exec in logso this isn't required

File: lib/modules/manager/pub/artifacts.ts [hunk 4]
```diff
@@ -60,16 +76,15 @@ export async function updateArtifacts({
...
+    };
 
+    await exec(cmd, execOptions);
     const newLockFileContent = await readLocalFile(lockFileName, 'utf8');
-    if (
-      oldLockFileContent === newLockFileContent ||
-      newLockFileContent === undefined
-    ) {
+    if (oldLockFileContent === newLockFileContent) {
+      logger.debug(`${lockFileName} is unchanged`);
```
Comment (viceice) [inline lib/modules/manager/pub/artifacts.ts:84]:
```suggestion
```
upper layer will log if there are updated log files

File: lib/modules/manager/pub/artifacts.ts [hunk 5]
```diff
@@ -60,16 +76,15 @@ export async function updateArtifacts({
...
       return null;
     }
-
+    logger.debug(`Returning updated ${lockFileName}`);
```
Comment (viceice) [inline lib/modules/manager/pub/artifacts.ts:87]:
```suggestion
```
see other comment

**Review by viceice — CHANGES_REQUESTED** [review]

**Review by JamieMagee — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**